docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,505 @@
+  Class string `json:"class,omitempty"`
+
+
+  // Template is a local struct containing a collection of templates for creation of
+  // MachineDeployment objects representing a set of worker nodes.
+  Template TemplateClass `json:"template"`
+}
+
+type TemplateClass struct {
+  Metadata ObjectMeta `json:"metadata,omitempty"`
+
+  // Bootstrap contains the bootstrap template references to be used
+  // for the creation of worker Machines.
+  Bootstrap LocalObjectTemplate `json:"bootstrap,omitempty"`
+
+  // Infrastructure contains the infrastructure template references to be used
+  // for the creation of worker Machines.
+  Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
+}
+
+type LocalObjectTemplate struct {
+  // Ref is a required reference to a custom resource
+  // offered by a bootstrap provider.
+  Ref corev1.ObjectReference `json:"ref"`
+}
+```
+
+#### Modification to existing API Types
+##### Cluster
+1.  Add `Cluster.Spec.Managed` defined as
+    ```golang
+    // The information of the managed resources for the cluster.
+    // +optional
+    Managed *ManagedCluster `json:"managed,omitempty"`
+    ```
+1.  The `ManagedCluster` object has the following definition:
+    ```golang
+    // ManagedCluster encapsulates the information of the managed resources.
+    type ManagedCluster struct {
+      // The name of the ClusterClass object to create the managed cluster topology.
+      Class string `json:"class,omitempty"`
+
+      // The kubernetes version of the cluster.
+      Version string `json:"version"`
+
+      // RolloutAfter performs a rollout of the entire cluster one component at a time [...]
+      // +optional
+      RolloutAfter *metav1.Time `json:"rolloutAfter,omitempty"`
+
+      // The information for the Control plane of the cluster.
+      ControlPlane ControlPlane `json:"controlPlane"`
+
+      // Worker encapsulates the different constructs that form the worker nodes
+      // for the cluster.
+      // +optional
+      Worker *Worker `json:"worker,omitempty"`
+    }
+    ```
+1.  The `ControlPlane` object contains the parameters for the control plane nodes of the managed cluster.
+    ```golang
+    // ControlPlane specifies the parameters for the control plane nodes in the cluster.
+    type ControlPlane struct {
+     Metadata ObjectMeta `json:"metadata,omitempty"`
+
+     // The number of control plane nodes.
+     Replicas int `json:"replicas"`
+    }
+    ```
+1.  The `Worker` object represents the sets of worker nodes of the managed cluster.
+    ```golang
+    // Worker represents the different sets of worker nodes in the cluster.
+    type Worker struct {
+      Deployments []Deployment `json:”deployments,omitempty”`
+    }
+    ```
+1.  The `Deployment` object represents a single set of worker nodes of the managed cluster.
+    ```golang
+    // Deployment specifies the different parameters for a set of worker nodes in the cluster.
+    type Deployment struct {
```

Comment:
We should probably avoid the term deployment to avoid confusion with appsv1.Deployment

Reply:
yes

Reply:
The idea behind naming it as `Deployment` was to map it to `MachineDeployment`. And in the future, introduce `Pool` which would correspond to `MachinePool`.